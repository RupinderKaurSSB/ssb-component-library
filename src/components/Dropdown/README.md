Dropdown
========

> A component that allows a user to select a value from a series of options.

### Usage

#### HTML

```html
<div class="ssb-dropdown">
    <label for="input-dropdown">Menu header</label>
    <div class="dropdown-interactive-area">
        <input class="focused" id="input-dropdown" disabled="" placeholder="Select item" value="">
        <div class="dd-icon"><i /></div>
        <div class="list-of-options">
            <div class="option-list-element" value="item1" id="item1">Apples</div>
            <div class="option-list-element selected" value="item2" id="item2">Rainbows</div>
            <div class="option-list-element" value="item3" id="item3">Ocean</div>
        </div>
    </div>
</div>
```

#### React

```jsx harmony
<Dropdown header="Menu header" icon={<ChevronDown />} items={items} />

<Dropdown header="Menu header" icon={<ChevronDown />} selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />

<Dropdown header="Menu header" icon={<ChevronDown />} searchable items={items} />
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| header | string | Displays header |
| icon | node | Renders an icon |
| items | arrayOf(title, id) |Required items for rendering dropdown |
| onSelect | func | Callback function when a title is selected |
| open | bool | Selected state of dropdown |
| placeholder | string | Placeholder for dropdown |
| searchable | bool | If true, dropdown is searchable |
| selectedItem | object | Selected element in dropdown |
