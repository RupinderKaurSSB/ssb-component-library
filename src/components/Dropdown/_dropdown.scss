.ssb-dropdown {
  @include roboto;
  display: flex;
  flex-direction: column;
  transition: all .25s ease-in-out;
  max-width: 350px;
  min-width: 150px;

  label {
    font-size: 14px;
    margin-bottom: 5px;
  }

  .dropdown-interactive-area {
    @include reset-button;
    cursor: pointer;
    position: relative;

    .dd-icon {
      color: $ssb-green-4;
      font-size: 18px;
      position: absolute;
      right: 10px;
      top: 10px;
    }

    input {
      background-color: $ssb-white;
      border: 1px solid $ssb-dark-6;
      box-sizing: border-box;
      cursor: pointer;
      font-size: 16px;
      height: 40px;
      padding: 4px 40px 4px 10px;
      text-overflow: ellipsis;
      width: 100%;

      &::placeholder {
        color: $ssb-dark-6;
      }

      &:hover {
        border: 1px solid $ssb-green-4;
      }

      &:focus, &.focused {
        border: 1px solid $ssb-green-4;
        outline: $ssb-green-4 auto 1px;
      }
    }

    .list-of-options {
      @include roboto;
      background: $ssb-white;
      border: 1px solid $ssb-dark-6;
      left: 0;
      list-style: none;
      margin: 0;
      max-height: 235px;
      min-width: 150px;
      overflow-y: auto;
      padding-left: 0;
      position: absolute;
      scrollbar-color: $ssb-dark-2 $ssb-white;
      scrollbar-width: thin;
      top: 40px;
      z-index: 9999;
      width: 100%;

      .option-list-element {
        @include reset-button;
        background: $ssb-white;
        cursor: pointer;
        display: block;
        overflow: hidden;
        padding: 8px 12px;
        text-overflow: ellipsis;
        transition: all .18s;
        width: 100%;

        &:hover, &:focus {
          background: $ssb-dark-5;
          color: $white;
          z-index: 2;
        }

        &.selected {
          background: $ssb-green-4;
          color: $ssb-white;
        }
      }

    }

    ::-webkit-scrollbar {
      width: 6px;
    }

    ::-webkit-scrollbar-track {
      background: $ssb-white;
      border-bottom: 1px solid $ssb-dark-6;
    }

    ::-webkit-scrollbar-thumb {
      background: $ssb-dark-2;
      border-radius: 3px;
    }

    ::-webkit-scrollbar-thumb:hover {
      background: $ssb-dark-4;
    }

  }
}
