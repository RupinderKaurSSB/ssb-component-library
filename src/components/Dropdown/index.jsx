import React, { useEffect, useRef, useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import uuid from 'uuid/v4';
import { ChevronDown } from 'react-feather';
import { ssbDark3, ssbDark4, ssbDark6, ssbGreen2, ssbGreen4, ssbWhite } from '../../style/colors';
import { roboto } from '../../style/mixins';

const Dropdown = ({
	header, items, onSelect, open, placeholder, searchable, selectedItem,
}) => {
	const id = uuid();
	const node = useRef();
	const [isOpen, setOpen] = useState(open);
	const [availableOptions, filterOptions] = useState(items);
	const [selectedOption, selectItem] = useState(selectedItem || { title: '', id: '' });
	const [inputFieldValue, updateInputValue] = useState('');

	const filterItems = event => {
		updateInputValue(event.target.value);
		filterOptions(items.filter(it => it.title.toLowerCase().includes(event.target.value.toLowerCase())));
	};

	const handleSelection = item => {
		selectItem({ title: item.title, id: item.id });
		onSelect(item);
		setOpen(false);
		filterOptions(items);
		updateInputValue('');
	};

	const handleClickOutside = e => {
		if (!node.current.contains(e.target)) {
			setOpen(false);
		}
	};

	useEffect(() => {
		if (isOpen) {
			document.addEventListener('mousedown', handleClickOutside);
		} else {
			document.removeEventListener('mousedown', handleClickOutside);
		}
		return () => {
			document.removeEventListener('mousedown', handleClickOutside);
		};
	}, [isOpen]);

	return (
		<DropdownWrapper>
			{header && <label htmlFor={id}>{header}</label>}
			<InteractiveArea className="dropdown-interactive-area" ref={node} onClick={() => setOpen(!isOpen)}>
					<InputWrapper
						className={isOpen ? 'focused' : ''}
						id={id}
						onChange={e => filterItems(e)}
						disabled={!searchable}
						placeholder={selectedOption.title ? selectedOption.title : placeholder}
						value={inputFieldValue}
					/>
					<ChevronDown className="chevron-icon" size={18} />
				{isOpen && (
					<OptionListWrapper>
						{availableOptions.map(it => (
							<li
								className={`option-list-element ${selectedOption.id === it.id && 'selected'}`}
								key={it.id}
							>
								<option
									onClick={() => handleSelection(it)}
									value={it.id}
									id={it.id}
								>{it.title}
								</option>
							</li>
						))}
					</OptionListWrapper>
				)}
			</InteractiveArea>
		</DropdownWrapper>
	);
};

const DropdownWrapper = styled.div`
	${roboto};
	display: flex;
	flex-direction: column;
	transition: all .25s ease-in-out;
  
	label {
		margin-bottom: 5px;
		font-size: 14px;
	}
  
	.dropdown-interactive-area {
		cursor: pointer;
		position: relative;        
	}
`;

const InteractiveArea = styled.div`
	cursor: pointer;
	position: relative;
	
	.chevron-icon {
		color: ${ssbGreen4};
		position: absolute;
		top: 10px;
		right: 10px;
	}
`;

const InputWrapper = styled.input`
	border: 1px solid ${ssbDark6};
	cursor: pointer;
	font-size: 16px;
	height: 36px;
	padding: 4px 34px 4px 10px;
	background-color: ${ssbWhite};
	width: 100%;

	&::placeholder {
		color: ${ssbDark6};
	}

	&:hover {
		border: 1px solid ${ssbGreen4};
	}

	&:focus, &.focused {
		border: 1px solid ${ssbGreen4};
		outline: ${ssbGreen4} auto 1px;
	}
`;

const OptionListWrapper = styled.ul`
	background: ${ssbWhite};
	border: 1px solid ${ssbDark6};
	font-family: 'Roboto', sans-serif;
	height: 235px;
	left: 0;
	list-style: none;
	margin: 0;
	overflow-y: scroll;
	padding-left: 0;
	position: absolute;
	scrollbar-color: ${ssbDark3} ${ssbWhite};
	scrollbar-width: thin;
	top: 36px;
	z-index: 9999;
	width: 100%;
	
	::-webkit-scrollbar {
		width: 6px;
	}
	
	::-webkit-scrollbar-track {
		background: ${ssbWhite}; 
	}
	
	::-webkit-scrollbar-thumb { 
		background: ${ssbDark3};
		border-radius: 3px;
	}
	
	::-webkit-scrollbar-thumb:hover {
		background: ${ssbDark4}; 
	}

	.option-list-element {
		cursor: pointer;

		option {
			overflow: hidden;
			padding: 10px;
			text-overflow: ellipsis;
		}

		&:hover {
			background: ${ssbGreen2};
		}

		&.selected {
			background: ${ssbGreen4};
			color: ${ssbWhite};
		}
	}
`;

Dropdown.defaultProps = {
	header: '',
	items: [{ id: '', title: '' }],
	onSelect: () => {},
	open: false,
	searchable: false,
	placeholder: '-- Select --',
};

Dropdown.propTypes = {
	header: PropTypes.string,
	items: PropTypes.arrayOf(PropTypes.shape({
		title: PropTypes.string,
		id: PropTypes.string,
	})),
	onSelect: PropTypes.func,
	open: PropTypes.bool,
	placeholder: PropTypes.string,
	searchable: PropTypes.bool,
	selectedItem: PropTypes.object,
};

export default Dropdown;
