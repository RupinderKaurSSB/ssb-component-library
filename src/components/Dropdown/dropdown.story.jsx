import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import Dropdown from './index';

const items = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	}, {
		title: 'Automobiles',
		id: 'item4',
	}, {
		title: 'Banan',
		id: 'item5',
	}, {
		title: 'Appelsin',
		id: 'item6',
	}, {
		title: 'Fersken',
		id: 'item7',
	}, {
		title: 'Mandarin',
		id: 'item8',
	},
];

storiesOf('Atoms|Dropdown', module).addDecorator(centered)
	.add('Standard', () => (
		<Dropdown header="Menu header" items={items} />
	))
	.add('Pre-selected value', () => (
		<Dropdown header="Menu header" selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />
	))
	.add('Searchable', () => (
		<Dropdown header="Menu header" searchable items={items} />
	));
