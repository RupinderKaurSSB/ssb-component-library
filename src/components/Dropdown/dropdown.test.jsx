import React from 'react';
import {shallow} from 'enzyme';
import Dropdown from './index';

const items = [
	{
		title: 'Apples',
		id: 'item1',
	}, {
		title: 'Rainbows',
		id: 'item2',
	}, {
		title: 'Ocean',
		id: 'item3',
	},
];

jest.mock('uuid/v4', () => jest.fn(() => 1));

describe('Dropdown component', () => {
	test('Matches the snapshot', () => {
		const wrapper = shallow(<Dropdown header="Menu header" items={items} />);
		expect(wrapper).toMatchSnapshot();
	});
	test('preselected value', () => {
		const wrapper = shallow(<Dropdown header="Menu header" selectedItem={{ title: 'Ocean', id: 'item3' }} items={items} />);
		expect(wrapper.find('.dropdown-interactive-area').find('input').props().placeholder).toEqual('Ocean');
	});
	test('Searchable dropdown', () => {
		const wrapper = shallow(<Dropdown header="Menu header" searchable items={items} placeholder="Search and select" />);
		expect(wrapper.find('.dropdown-interactive-area').find('input').props().disabled).toEqual(false);
	});
	test('Change classname to open on click', () => {
		const wrapper = shallow(<Dropdown header="Menu header" items={items} />);
		wrapper.find('.dropdown-interactive-area').simulate('click');
		expect(wrapper.find('input').hasClass('focused')).toEqual(true);
		expect(wrapper.exists('.list-of-options')).toEqual(true);
	});
	test('Choose first item', () => {
		const wrapper = shallow(<Dropdown header="Menu header" items={items} open />);
		wrapper.find('.list-of-options').find('button').first().simulate('click');
		expect(wrapper.find('.dropdown-interactive-area').find('input').props().placeholder).toEqual('Apples');
	});

});
