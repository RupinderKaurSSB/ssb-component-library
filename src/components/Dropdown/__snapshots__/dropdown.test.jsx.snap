// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component Matches the snapshot 1`] = `
<div
  className="ssb-dropdown"
>
  <label
    htmlFor={1}
  >
    Menu header
  </label>
  <div
    className="dropdown-interactive-area"
    onKeyDown={[Function]}
  >
    <button
      aria-label="open or close dropdown"
      onClick={[Function]}
      onKeyDown={[Function]}
    >
      <input
        className=""
        disabled={true}
        id={1}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="-- Select --"
        value=""
      />
      <ChevronDown
        className="dd-icon"
        color="currentColor"
        size={24}
      />
    </button>
  </div>
</div>
`;
